File: Section 04/Section 04_SVM_n_NN.py
import pandas as pd
letterdata = pd.read_csv("letterdata.csv")
print (letterdata.head())

x_vars = letterdata.drop(['letter'],axis=1)
y_var = letterdata["letter"]

y_var = y_var.replace({'A':1,'B':2,'C':3,'D':4,'E':5,'F':6,'G':7,'H':8,'I':9,'J':10,
'K':11,'L':12,'M':13,'N':14,'O':15,'P':16,'Q':17,'R':18,'S':19,'T':20,
'U':21,'V':22,'W':23,'X':24,'Y':25,'Z':26})

from sklearn.metrics import accuracy_score,classification_report
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x_vars,y_var,train_size = 0.7,random_state=42)


# Linear Classifier
from sklearn.svm import SVC
svm_fit = SVC(kernel='linear',C=1.0,random_state=43)
svm_fit.fit(x_train,y_train)

print ("\nSVM Linear Classifier - Train Confusion Matrix\n\n",pd.crosstab(y_train,svm_fit.predict(x_train),rownames = ["Actuall"],colnames = ["Predicted"]) )     
print ("\nSVM Linear Classifier - Train accuracy:",round(accuracy_score(y_train,svm_fit.predict(x_train)),3))
print ("\nSVM Linear Classifier - Train Classification Report\n",classification_report(y_train,svm_fit.predict(x_train)))

print ("\n\nSVM Linear Classifier - Test Confusion Matrix\n\n",pd.crosstab(y_test,svm_fit.predict(x_test),rownames = ["Actuall"],colnames = ["Predicted"]))      
print ("\nSVM Linear Classifier - Test accuracy:",round(accuracy_score(y_test,svm_fit.predict(x_test)),3))
print ("\nSVM Linear Classifier - Test Classification Report\n",classification_report(y_test,svm_fit.predict(x_test)))


#Polynomial Kernel
svm_poly_fit = SVC(kernel='poly',C=1.0,degree=2)
svm_poly_fit.fit(x_train,y_train)

print ("\nSVM Polynomial Kernel Classifier - Train Confusion Matrix\n\n",pd.crosstab(y_train,svm_poly_fit.predict(x_train),rownames = ["Actuall"],colnames = ["Predicted"]) )     
print ("\nSVM Polynomial Kernel Classifier - Train accuracy:",round(accuracy_score(y_train,svm_poly_fit.predict(x_train)),3))
print ("\nSVM Polynomial Kernel Classifier - Train Classification Report\n",classification_report(y_train,svm_poly_fit.predict(x_train)))

print ("\n\nSVM Polynomial Kernel Classifier - Test Confusion Matrix\n\n",pd.crosstab(y_test,svm_poly_fit.predict(x_test),rownames = ["Actuall"],colnames = ["Predicted"]))      
print ("\nSVM Polynomial Kernel Classifier - Test accuracy:",round(accuracy_score(y_test,svm_poly_fit.predict(x_test)),3))
print ("\nSVM Polynomial Kernel Classifier - Test Classification Report\n",classification_report(y_test,svm_poly_fit.predict(x_test)))


#RBF Kernel
svm_rbf_fit = SVC(kernel='rbf',C=1.0, gamma=0.1)
svm_rbf_fit.fit(x_train,y_train)

print ("\nSVM RBF Kernel Classifier - Train Confusion Matrix\n\n",pd.crosstab(y_train,svm_rbf_fit.predict(x_train),rownames = ["Actuall"],colnames = ["Predicted"]) )     
print ("\nSVM RBF Kernel Classifier - Train accuracy:",round(accuracy_score(y_train,svm_rbf_fit.predict(x_train)),3))
print ("\nSVM RBF Kernel Classifier - Train Classification Report\n",classification_report(y_train,svm_rbf_fit.predict(x_train)))

print ("\n\nSVM RBF Kernel Classifier - Test Confusion Matrix\n\n",pd.crosstab(y_test,svm_rbf_fit.predict(x_test),rownames = ["Actuall"],colnames = ["Predicted"]))      
print ("\nSVM RBF Kernel Classifier - Test accuracy:",round(accuracy_score(y_test,svm_rbf_fit.predict(x_test)),3))
print ("\nSVM RBF Kernel Classifier - Test Classification Report\n",classification_report(y_test,svm_rbf_fit.predict(x_test)))



# Grid Search - RBF Kernel
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split,GridSearchCV

pipeline = Pipeline([('clf',SVC(kernel='rbf',C=1,gamma=0.1 ))])

parameters = {'clf__C':(0.1,0.3,1,3,10,30),
              'clf__gamma':(0.001,0.01,0.1,0.3,1)}

grid_search_rbf = GridSearchCV(pipeline,parameters,n_jobs=-1,cv=5,verbose=1,scoring='accuracy')
grid_search_rbf.fit(x_train,y_train)


print ('RBF Kernel Grid Search Best Training score: %0.3f' % grid_search_rbf.best_score_)
print ('RBF Kernel Grid Search Best parameters set:')
best_parameters = grid_search_rbf.best_estimator_.get_params()
for param_name in sorted(parameters.keys()):
    print ('\t%s: %r' % (param_name, best_parameters[param_name]))

predictions = grid_search_rbf.predict(x_test)

print ("\nRBF Kernel Grid Search - Testing accuracy:",round(accuracy_score(y_test, predictions),4))
print ("\nRBF Kernel Grid Search - Test Classification Report\n",classification_report(y_test, predictions))
print ("\n\nRBF Kernel Grid Search- Test Confusion Matrix\n\n",pd.crosstab(y_test, predictions,rownames = ["Actuall"],colnames = ["Predicted"]))      

File: Section 04/Section 04_SVM_n_NN.R
letter_data = read.csv("letterdata.csv")

set.seed(123)
numrow = nrow(letter_data)
trnind = sample(1:numrow,size = as.integer(0.7*numrow))
train_data = letter_data[trnind,]
test_data = letter_data[-trnind,]


library(e1071)

accrcy <- function(matrx){ 
  return( sum(diag(matrx)/sum(matrx)))}

precsn <- function(matrx){
  return(diag(matrx) / rowSums(matrx))
}

recll <- function(matrx){
  return(diag(matrx) / colSums(matrx))
}




# SVM - Linear Kernel
svm_fit = svm(letter~.,data = train_data,kernel="linear",cost=1.0,scale = TRUE)

tr_y_pred = predict(svm_fit, train_data)
ts_y_pred = predict(svm_fit,test_data)

tr_y_act = train_data$letter;ts_y_act = test_data$letter

tr_tble = table(tr_y_act,tr_y_pred)
print(paste("Train Confusion Matrix"))
print(tr_tble)
tr_acc = accrcy(tr_tble)
print(paste("SVM Linear Kernel Train accuracy:",round(tr_acc,4)))

tr_prec = precsn(tr_tble)
print(paste("SVM Linear Kernel Train Precision:"))
print(tr_prec)

tr_rcl = recll(tr_tble)
print(paste("SVM Linear Kernel Train Recall:"))
print(tr_rcl)

ts_tble = table(ts_y_act,ts_y_pred)
print(paste("Test Confusion Matrix"))
print(ts_tble)

ts_acc = accrcy(ts_tble)
print(paste("SVM Linear Kernel Test accuracy:",round(ts_acc,4)))

ts_prec = precsn(ts_tble)
print(paste("SVM Linear Kernel Test Precision:"))
print(ts_prec)

ts_rcl = recll(ts_tble)
print(paste("SVM Linear Kernel Test Recall:"))
print(ts_rcl)


# SVM - Polynomial Kernel
svm_poly_fit = svm(letter~.,data = train_data,kernel="poly",cost=1.0,degree = 2  ,scale = TRUE)

tr_y_pred = predict(svm_poly_fit, train_data)
ts_y_pred = predict(svm_poly_fit,test_data)

tr_y_act = train_data$letter;ts_y_act = test_data$letter


tr_tble = table(tr_y_act,tr_y_pred)
print(paste("Train Confusion Matrix"))
print(tr_tble)
tr_acc = accrcy(tr_tble)
print(paste("SVM Polynomial Kernel Train accuracy:",round(tr_acc,4)))

tr_prec = precsn(tr_tble)
print(paste("SVM Polynomial Kernel Train Precision:"))
print(tr_prec)

tr_rcl = recll(tr_tble)
print(paste("SVM Polynomial Kernel Train Recall:"))
print(tr_rcl)

ts_tble = table(ts_y_act,ts_y_pred)
print(paste("Test Confusion Matrix"))
print(ts_tble)

ts_acc = accrcy(ts_tble)
print(paste("SVM Polynomial Kernel Test accuracy:",round(ts_acc,4)))

ts_prec = precsn(ts_tble)
print(paste("SVM Polynomial Kernel Test Precision:"))
print(ts_prec)

ts_rcl = recll(ts_tble)
print(paste("SVM Polynomial Kernel Test Recall:"))
print(ts_rcl)



# SVM - RBF Kernel
svm_rbf_fit = svm(letter~.,data = train_data,kernel="radial",cost=1.0,gamma = 0.2  ,scale = TRUE)

tr_y_pred = predict(svm_rbf_fit, train_data)
ts_y_pred = predict(svm_rbf_fit,test_data)

tr_y_act = train_data$letter;ts_y_act = test_data$letter

tr_tble = table(tr_y_act,tr_y_pred)
print(paste("Train Confusion Matrix"))
print(tr_tble)
tr_acc = accrcy(tr_tble)
print(paste("SVM RBF Kernel Train accuracy:",round(tr_acc,4)))

tr_prec = precsn(tr_tble)
print(paste("SVM RBF Kernel Train Precision:"))
print(tr_prec)

tr_rcl = recll(tr_tble)
print(paste("SVM RBF Kernel Train Recall:"))
print(tr_rcl)

ts_tble = table(ts_y_act,ts_y_pred)
print(paste("Test Confusion Matrix"))
print(ts_tble)

ts_acc = accrcy(ts_tble)
print(paste("SVM RBF Kernel Test accuracy:",round(ts_acc,4)))

ts_prec = precsn(ts_tble)
print(paste("SVM RBF Kernel Test Precision:"))
print(ts_prec)

ts_rcl = recll(ts_tble)
print(paste("SVM RBF Kernel Test Recall:"))
print(ts_rcl)



# Grid search - RBF Kernel
library(e1071)
svm_rbf_grid = tune(svm,letter~.,data = train_data,kernel="radial",scale=TRUE,ranges = list(
  cost = c(0.1,0.3,1,3,10,30),
  gamma = c(0.001,0.01,0.1,0.3,1)
  
),
tunecontrol = tune.control(cross = 5)
)

print(paste("Best parameter from Grid Search"))
print(summary(svm_rbf_grid))

best_model = svm_rbf_grid$best.model

tr_y_pred = predict(best_model,data = train_data,type = "response")
ts_y_pred = predict(best_model,newdata = test_data,type = "response")

tr_y_act = train_data$letter;ts_y_act = test_data$letter


tr_tble = table(tr_y_act,tr_y_pred)
print(paste("Train Confusion Matrix"))
print(tr_tble)
tr_acc = accrcy(tr_tble)
print(paste("SVM RBF Kernel Train accuracy:",round(tr_acc,4)))

tr_prec = precsn(tr_tble)
print(paste("SVM RBF Kernel Train Precision:"))
print(tr_prec)

tr_rcl = recll(tr_tble)
print(paste("SVM RBF Kernel Train Recall:"))
print(tr_rcl)

ts_tble = table(ts_y_act,ts_y_pred)
print(paste("Test Confusion Matrix"))
print(ts_tble)

ts_acc = accrcy(ts_tble)
print(paste("SVM RBF Kernel Test accuracy:",round(ts_acc,4)))

ts_prec = precsn(ts_tble)
print(paste("SVM RBF Kernel Test Precision:"))
print(ts_prec)

ts_rcl = recll(ts_tble)
print(paste("SVM RBF Kernel Test Recall:"))
print(ts_rcl)